src/libraries/System.Private.CoreLib/src/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs

```diff
@@ -135,13 +135,21 @@ public override T[] Rent(int minimumLength)
                 }
 
                 // No buffer available.  Allocate a new buffer with a size corresponding to the appropriate bucket.
-                buffer = GC.AllocateUninitializedArray<T>(_bucketArraySizes[bucketIndex]);
+#if !MONO
+                buffer = GC.AllocateUninitializedArray<T>(_bucketArraySizes[bucketIndex], pinned: !RuntimeHelpers.IsReferenceOrContainsReferences<T>());
+#else
+                buffer = GC.AllocateUninitializedArray<T>(_bucketArraySizes[bucketIndex], pinned: false);
+#endif
             }
             else
             {
                 // The request was for a size too large for the pool.  Allocate an array of exactly the requested length.
                 // When it's returned to the pool, we'll simply throw it away.
-                buffer = GC.AllocateUninitializedArray<T>(minimumLength);
+#if !MONO
+                buffer = GC.AllocateUninitializedArray<T>(minimumLength, pinned: !RuntimeHelpers.IsReferenceOrContainsReferences<T>());
+#else
+                buffer = GC.AllocateUninitializedArray<T>(minimumLength, pinned: false);
+#endif
```

Comment:
Isn't it counter-productive to pin the arrays that will be thrown away when returned to the pool?

Reply:
Here "pinnig" is just allocating into a separate region of the heap that is never compacted. It does not mean some additional action.
In this particular case the array is very large (2^31 elements ?). If we do not specify "pinned" it will end up in the Large Object Heap, which is very similar to Pinned Object Heap, just permits compaction, which may not be a good idea with such a large object.

The main reason to use "pinned" here is consistency with other cases.

Reply:
Wouldn't the limit be 2^21?

Reply:
Yes 2^21 
(4+17), to be sure. :-)
It is a big array.